File: 0329-longest-increasing-path-in-a-matrix/0329-longest-increasing-path-in-a-matrix.cpp
class Solution {
public:
    
    int memo[40001];
    
    int twoToOne(pair<int,int> p, int m, int n)
    {
        return p.first * n + p.second;
    }
    
    int dfs(int curt, vector<vector<int>>& adj)
    {
        if(memo[curt] != 0)
        {
            return memo[curt];
        }
        vector<int> neighbors = adj[curt];
        
        for(auto neighbor : neighbors)
        {
            memo[curt] = max(memo[curt], dfs(neighbor, adj));
        }
        
        
        return ++memo[curt];
    }
    
    int longestIncreasingPath(vector<vector<int>>& matrix) {
        int m = matrix.size(), n = matrix[0].size();
        
        vector<vector<int>> adj(m * n, vector<int>(0));
        unordered_map<int,int> map;
        
        for(int i = 0; i < m; i++)
        {
            for(int j = 0; j < n; j++)
            {
                map[twoToOne({i, j}, m, n)] = 0;
            }
        }
        
        int directions[4][2] = {{1,0},{-1,0},{0,1},{0,-1}};
        
        for(int x = 0; x < m; x++)
        {
            for(int y = 0; y < n; y++)
            {
                
                for(auto& direction : directions)
                {
                    int newx = x + direction[0];
                    int newy = y + direction[1];
                    
                    if(newx < 0 || newx >= m || newy < 0 || newy >= n)
                    {
                        continue;
                    }
                    
                    if(matrix[newx][newy] <= matrix[x][y]) 
                    {
                        continue;
                    }
                    
                    int curt = twoToOne({x, y}, m, n);
                    int next = twoToOne({newx, newy}, m, n);
                    
                    adj[curt].push_back(next);
                    map[next]++;
                }
            }
        }
        
        int ans = 0;
        
        for(auto& [key, value] : map)
        {
            if(value == 0)
                ans = max(ans, dfs(key, adj));
        }
        
        return ans;
    }
};

File: 0329-longest-increasing-path-in-a-matrix/0329-longest-increasing-path-in-a-matrix.py
class Solution:
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        m, n = len(matrix), len(matrix[0])
        
        dp = [[-1 for _ in range(n)] for _ in range(m)]
        ans = 0
        
        def dfs(i, j):
            if dp[i][j] != -1:
                return dp[i][j]
            
            used = 0
            
            for direction in [[1,0],[-1,0],[0,1],[0,-1]]:
                newx, newy = i + direction[0], j + direction[1]
                
                if newx < 0 or newx >= len(matrix) or newy < 0 or newy >= len(matrix[0]):
                    continue
                
                if matrix[newx][newy] <= matrix[i][j]:
                    continue
                    
                used = max(used, dfs(newx, newy))
            
            dp[i][j] = used + 1
            return dp[i][j]
        
        for i in range(m):
            for j in range(n):
                ans = max(ans, dfs(i, j))
        
        return ans
        
                
                